### ISSUE TITLE
Improve representation of dashed lines according to ISOM2017-2

### ISSUE THREAD

**Issue by valdisj** [OpenOrienteering/mapper#1936]
Currently dashed symbols in MAPPER are drawn using minimal length according to ISOM2017-2 as per symbol spec. Though, ISOM2017-2 allows dashed symbols to be shorter by coeficient of x0.8. This results in a situation, that other software used for ISOM 2017-2 puts this condition in use, and often mappers draw them slightly shorter than minimum lenght as per symbol spec, but still according to ISOM 2017-2 spec, and such map, when opened in OOM, shows wrong representation of this symbol.
![image](https://user-images.githubusercontent.com/5184678/114160356-7fb4cc80-992f-11eb-8b00-64768841228c.png)
Suggestion is to implement this coeficient of x 0.8 for dashed symbols throughout the OpenOrienteering Mapper, as per ISOM 2017-2 specs.

**Comment by dg0yt** [comment]
Thank you very much. However, I think it needs to be interpreted in a different way.
- **The minimum symbol size (lenght of the line) must not be reduced.**
- If the length of the line is not an exact composition of regular dashes and gaps (which is the *normal* case), then the length of elements must be adjusted (to avoid a gap or excess length at the end). **In this situation,**
  -  gaps must not be adjusted,
  -  the dashes shall never be shorter than 0.8 times the regular length.

This does not allow for "slightly shorter than minimum lenght as per symbol spec".

**Comment by valdisj** [comment]
I agree on GAPs. That is confirmed. But on the dashes, Mapper acts slightly differently. This is the result of actuals:
<img width="614" alt="image" src="https://user-images.githubusercontent.com/5184678/114172710-af1f0580-993e-11eb-9e76-3b543d606102.png">
[less distinct trail.omap.zip](https://github.com/OpenOrienteering/mapper/files/6285356/less.distinct.trail.omap.zip)

**Comment by krticka** [comment]
I am sometimes facing the same problem described by @valdisj when mapping. This led me to search old emails when we discussed the request to define gaps and dash lengths for ISOM 2017. Of course there must be some variability allowed to be able to efficiency draw styled lines. Several variants were proposed by Håvard Tveite and we agreed on solution: Dash lengths are allowed to vary. Dash lengths must be within some percentage (+/-20%) of the length given in the map specification.
Maybe we were not enough clear in the specification. But longer dashes were seen as no problem, so we used the only phrase "never shorter than 0.8 times the given length". 
I think that allow to draw symbol 507 from 0.8 mm dashes would help many mappers to use this symbol. Than the min. size from which symbol would render double dashes would be 4.5 mm / 67.5 m.

**Comment by dg0yt** [comment]
> Dash lengths must be within some percentage (+/-20%) of the length given in the map specification.

In particular for short lines, the +/- 20% is too strict, leaving some line length intervalls without solution:
![Unbenannt](https://user-images.githubusercontent.com/13567791/114237226-0d6dd780-9983-11eb-9d99-dcaaf31e350c.png)

> I think that allow to draw symbol 507 from 0.8 mm dashes would help many mappers to use this symbol. 

But this needs to be made official. Mappers can help themselves anyway and change symbol definitions. Or suggest/contribute changes in our interpretation/implementation here.

**Comment by krticka** [comment]
> In particular for short lines, the +/- 20% is too strict, leaving some line length intervalls without solution:
I understand, that's maybe the reason why we mentioned only lower limit in the specification. Can't remember exactly the final discussion about this at MC meeting, it is five years back. Longer dashes are not the issue, too short ones are the problem. We we can leave specification as it is.

> Mappers can help themselves anyway and change symbol definitions. 
Sure. If I change the min. length line to 4.5 mm, the double dashes are created at distance 4.88 mm/ 73 m and the dash is 0.88 mm. What am I doing wrong?

**Comment by dg0yt** [comment]
> What am I doing wrong?

I don't think you are doing anything wrong at this point. 

In the past, reports about the drawing of short line all focused on what could be called abuse, requesting what was possible in other software. This time it is different, focusing on valid shortcomings in the current implementation with regard to the specification. This is positive.

The dash pattern generation in Mapper was never changed. The downside is that it doesn't match above specification. The upside is that we never broke existing maps or existing interoperability with other software as Purple Pen. When we create new mode of generating dash patterns, it needs a new file format version. In particular for line symbols, I think this needs a different approach anyway, a sort of formal language.

**Comment by valdisj** [comment]
Thanks @krticka and @dg0yt for getting into this and seeing a problem. I believe, if IOF Mapping commission would in next iteration clarify min/max and gaps in the ISOM2017-2 specs, this would also push developers of other software to improve their interpretation, which is, frankly, quite aggressively outside ISOM2017-2 requirements, at least for now. 
But as regards to OO Mapper, some adjustments proposed by @krticka would really make sense. I, personally, hate to change template to make workarounds. I'm very happy about strict rules which OO Mapper enforces and this helps to maintain quality. Importing/Exporting wrongly interpreted and therefore wrongly drawn shorter 507 and also 508 makes a real problem. And this is an issue large amount of people in orienteering community has now. But it all starts with clear documents and specs, leaving as minimal as possible interpretation freedom.
So, what are the solutions to the issue:
1) IOF mapping community needs to step-up and adjust (make more precise) requirements and then enforce adjusted specs this to all "out of the box" **templates** coming from all software producers. Also within our OO Mapper community.
2) More flexible adjustments in 507 representation would be very welcome in OO Mapper sw.

**Comment by krticka** [comment]
> The dash pattern generation in Mapper was never changed. The downside is that it doesn't match above specification. The upside is that we never broke existing maps or existing interoperability with other software as Purple Pen. When we create new mode of generating dash patterns, it needs a new file format version. In particular for line symbols, I think this needs a different approach anyway, a sort of formal language.

@dg0yt Breaking current maps and interoperability would certainly be a step in the wrong direction. For now I would support keep current software implementation but adjust dashed symbols to the extent which is currently possible.

@valdisj We defined this for the first time in ISOM 2017. Not sure if we want to be more strict. We can define everything more precisely, but from my experience, some "pillow" for implementation is useful. 

> enforce adjusted specs this to all "out of the box" templates coming from all software producers.

MC is the body which cares about development of mapping standards and map quality for major IOF events. MC does not have a legal status to force any software developer to change anything. MC can point out bugs and it is definitely in the software developer's own interest to fix the bugs, otherwise he could have dissatisfied customers/map users.

**Comment by valdisj** [comment]
I agree.

> > The dash pattern generation in Mapper was never changed. The downside is that it doesn't match above specification. The upside is that we never broke existing maps or existing interoperability with other software as Purple Pen. When we create new mode of generating dash patterns, it needs a new file format version. In particular for line symbols, I think this needs a different approach anyway, a sort of formal language.
> 
> @dg0yt Breaking current maps and interoperability would certainly be a step in the wrong direction. For now I would support keep current software implementation but adjust dashed symbols to the extent which is currently possible.
> 
> @valdisj We defined this for the first time in ISOM 2017. Not sure if we want to be more strict. We can define everything more precisely, but from my experience, some "pillow" for implementation is useful.
> 
> > enforce adjusted specs this to all "out of the box" templates coming from all software producers.
> 
> MC is the body which cares about development of mapping standards and map quality for major IOF events. MC does not have a legal status to force any software developer to change anything. MC can point out bugs and it is definitely in the software developer's own interest to fix the bugs, otherwise he could have dissatisfied customers/map users.
